server/src/main/java/io/druid/server/router/ConsistentHasher.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.router;
+
+import com.google.common.base.Charsets;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectRBTreeMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectSortedMap;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Distributes objects across a set of node keys using consistent hashing.
+ * See https://en.wikipedia.org/wiki/Consistent_hashing
+ * Not thread-safe.
+ */
+public class ConsistentHasher
+{
+  private static final int REPLICATION_FACTOR = 128;
+  private static final HashFunction DEFAULT_HASH_FN = Hashing.murmur3_128();
+
+  private final Long2ObjectRBTreeMap<String> nodeKeySlots = new Long2ObjectRBTreeMap<>();
+  {
+    nodeKeySlots.defaultReturnValue(null);
+  }
+  private final HashFunction hashFn;
+  private final Map<String, long[]> nodeKeyHashes = new HashMap<>();
+  private Set<String> previousKeys = new HashSet<>();
+
+  public ConsistentHasher(
+      final HashFunction hashFunction
+  )
+  {
+    this.hashFn = hashFunction == null ? DEFAULT_HASH_FN : hashFunction;
+  }
+
+  public void updateKeys(Set<String> currentKeys)
+  {
+    Set<String> added = new HashSet<>(currentKeys);
+    added.removeAll(previousKeys);
+
+    Set<String> removed = new HashSet<>(previousKeys);
+    removed.removeAll(currentKeys);
+
+    for (String key : added) {
+      addKey(key);
+    }
+
+    for (String key : removed) {
+      removeKey(key);
+    }
+
+    // store a copy in case the input was immutable
+    previousKeys = new HashSet<>(currentKeys);
+  }
+
+  public String hash(byte[] obj)
```

Comment:
a `hash(byte[])` returning String key sounds un-intuitive, can we call it something like `findKey(..)` ?

Reply:
renamed to findKey